Misleading steps in work with development container

troy0820 opened the pull request:
## Open Source Documentation Change 

#### When trying to contribute to Docker, the steps listed in the docs were misleading.  I couldn't get my changes from my local editor in the container with the way the docs are set up now.  

#### In Task 3 it states 
```
Save and close the cli/command/container/attach.go file.
Go to your running development container.
Remake the binary and copy it to usr/bin
Restart the Docker daemon with the new binary.
```

With these steps it doesn't tell you that you have to `make shell` again to push your local changes to a newly created image so you can continue to follow the steps.  

These changes add the step to get a new shell with the changes you made locally and then `hack/make.sh binary` to build the binaries to copy them into the "containers" /usr/bin.

With this change, hopefully it'll give people who may be confused the direction they need to continue to contribute to docker.  

Also the version that is now current in the branch clone from the `docker/docker` repository is 1.13.0-dev, which is changed and reflected in this pull request.

#### Cute cat image because it doesn't hurt
![cute-cat](https://cloud.githubusercontent.com/assets/5985850/19487278/1a043af4-9530-11e6-8fc5-79c29531fcf0.jpeg)

tophj-ibm commented:
The docs were misleading because this pr from the old docker/opensource repo wasn't moved over https://github.com/docker/opensource/pull/90/files.

troy0820 commented:
After discussion of the commands in the docker container vs the localhost shell, suggestions from this issue, came to this conclusion: 


After a code change, you can still make the binaries with hack/make.sh binary and then copy the contents to the /usr/bin with cp bundles/1.13.0-dev/binary-daemon/docker* /usr/bin && cp bundles/1.13.0-dev/binary-client/docker* /usr/bin in the docker container shell which the docs tell you to do but the make command isn't clearly distinguished from the commands listed above which has you do that to initially make the docker development container.

So in between these steps :

* Remake the binary and copy it to usr/bin
* Restart the Docker daemon with the new binary.
The make command should explicitly state the make command that the docker container uses hack/make.sh binary and the following instructions to copy the binaries to the /usr/bin in the docker container

Also, you shouldn't have to restart the docker daemon after the binaries are copied to the /usr/bin/ of the docker dev container because the commands you executed are in the docker dev container.


This is from the issue [docker-issue](https://github.com/docker/docker/issues/27495)

Question concerning this PR is:
* Update the version in the docs?
* Clarify the commands you are running in the docker dev container from above
* We shouldn't have to restart the docker daemon if we are entering these commands from the docker container after copying the binaries to the /usr/bin

This PR can be updated to reflect the changes to have the docs say the explicit commands from the docker dev container, updating the version, and removing the restart docker daemon command.

johndmulhausen commented:
@troy0820 @tophj-ibm @mstanleyjones I have a feeling that PR #90 on the old opensource repo was in fact pulled over, but it was added to https://github.com/docker/docker.github.io/tree/vnext-opensource

If you are making docs changes for the next version of Docker (which this PR seems to be, judging by all the references to 1.13), you should be working in the `vnext-` branch for that project. When that project's release is out, we merge it into master, which can also be thought of as the "live" version of the docs.

troy0820 commented:
@johndmulhausen Which ever is easier.  If updating these docs will migrate to the `vnext-opensource` branch then I'll change the `1.13.0-dev` back to `1.12.0-dev` to get the changes for the contributing back up there immediately and worry about the `1.13.0-dev` in the `vnext-opensource` branch.

johndmulhausen commented:
@troy0820 Yep, we 1.13.0-dev work should be documented in `vnext-opensource`, sorry for the confusion. So those parts should be reverted to 1.12.0-dev in this PR, and in a separate PR against `vnext-opensource` you would have the change to `1.13.0-dev`

Of course, we do want to retain your modifications to step 6, which instructs people on `make shell` -- that can stay here, in this PR. When we publish `vnext-opensource`'s contents with the release of the 1.13 Engine, the modifications to step 6 you make in master (in this PR) will be retained. :)

Make sense?

troy0820 commented:
@johndmulhausen yes, makes perfect sense. I'll rebase and make the necessary changes in this repo and update the other branch and make a PR to that.

troy0820 commented:
Okay @johndmulhausen I rebased but the changes I made are as follows:

* Rebased from docker/docker.github.io/master 
* Allow the user to set up docker development container by using the make command `make BIN_DIR=. shell` to bind the entire repo to the user's localhost.
* Allow the user to enter the make commands from the docker development container and copy the binaries from the docker dev container shell _instead of make shell after every edit from the localhost terminal as previously discussed_.
* Reverted back to version `1.12.0-dev`
* Deleted the part to restart the daemon after copying the binaries, as it isn't necessary from the docker development container shell.

mstanleyjones commented on opensource/project/set-up-dev-env.md:
```diff
@@ -275,13 +275,17 @@ example, you'll edit the help for the `attach` subcommand.
+6. Go to your running docker development container shell.
 
-7. Remake the binary and copy it to `usr/bin`
+7. Remake the binary by entering `hack/make.sh binary`  in the docker development container shell.
```
Consider "rebuild" or another verb. Maybe something like "using the command" instead of "entering".

Nit: double space before 'in'

mstanleyjones commented on opensource/project/set-up-dev-env.md:
````diff
@@ -275,13 +275,17 @@ example, you'll edit the help for the `attach` subcommand.
+```
 
-9. View your change by display the `attach` help.
+9. View your change by displaying the `attach` help in the docker development container shell.
````
Consider "To view your change, run the `docker help attach` command in the..."

mstanleyjones commented on opensource/project/set-up-dev-env.md:
````diff
@@ -275,13 +275,17 @@ example, you'll edit the help for the `attach` subcommand.
-9. View your change by display the `attach` help.
+9. View your change by displaying the `attach` help in the docker development container shell.
 
    ```none
````
This should be bash highlighting, I think...

troy0820 commented:
@mstanleyjones I fixed the changes from the code review but i didn't want more than one commit.  So I squashed the commit and forced push to have the one commit but with the changes you requested.

troy0820 commented:
Thank you, and hope to contribute some more. 👍

johndmulhausen commented:
Thanks so much @troy0820 ! :)
